File: backend/src/models/Volunteer.ts
import mongoose, { Schema, type Document } from 'mongoose';

export type VolunteerAvailability = 'AVAILABLE' | 'BUSY' | 'OFFLINE';

export interface IVolunteer extends Document {
  userId: mongoose.Types.ObjectId;
  skills: string[];
  location?: {
    type: 'Point';
    coordinates: [number, number];
  };
  availability: VolunteerAvailability;
  verificationStatus: 'PENDING' | 'VERIFIED' | 'REJECTED';
  currentAssignmentId?: mongoose.Types.ObjectId;
  experience?: number;
  createdAt: Date;
  updatedAt: Date;
}

const volunteerSchema = new Schema<IVolunteer>(
  {
    userId: { type: Schema.Types.ObjectId, ref: 'User', required: true, unique: true },
    skills: { type: [String], default: [] },
    location: {
      type: {
        type: String,
        enum: ['Point'],
        default: 'Point'
      },
      coordinates: { type: [Number] }
    },
    availability: {
      type: String,
      enum: ['AVAILABLE', 'BUSY', 'OFFLINE'],
      default: 'AVAILABLE'
    },
    verificationStatus: {
      type: String,
      enum: ['PENDING', 'VERIFIED', 'REJECTED'],
      default: 'PENDING'
    },
    currentAssignmentId: { type: Schema.Types.ObjectId, ref: 'Assignment' },
    experience: { type: Number, default: 0 }
  },
  { timestamps: true }
);

volunteerSchema.index({ location: '2dsphere' });
volunteerSchema.index({ availability: 1 });
volunteerSchema.index({ skills: 1 });

export const Volunteer = mongoose.model<IVolunteer>('Volunteer', volunteerSchema);

File: backend/src/main/java/com/raava/model/Volunteer.java
package com.raava.model;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.springframework.data.annotation.Id;
import org.springframework.data.mongodb.core.index.GeoSpatialIndexType;
import org.springframework.data.mongodb.core.index.GeoSpatialIndexed;
import org.springframework.data.mongodb.core.index.Indexed;
import org.springframework.data.mongodb.core.mapping.DBRef;
import org.springframework.data.mongodb.core.mapping.Document;

import java.time.Instant;
import java.util.ArrayList;
import java.util.List;

@Document(collection = "volunteers")
@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Volunteer {
    @Id
    private String id;

    @DBRef(lazy = true)
    private User user;

    @Builder.Default
    private List<String> skills = new ArrayList<>();

    @GeoSpatialIndexed(type = GeoSpatialIndexType.GEO_2D)
    private double[] location;

    private Availability availability = Availability.AVAILABLE;

    private VerificationStatus verificationStatus = VerificationStatus.PENDING;

    @DBRef(lazy = true)
    private Assignment currentAssignment;

    private int experience;

    @Indexed
    private Instant createdAt;

    @Indexed
    private Instant updatedAt;

    public enum Availability { AVAILABLE, BUSY, OFFLINE }
    public enum VerificationStatus { PENDING, VERIFIED, REJECTED }
}
